### PR TITLE
[Feature] recover partition support lake table

### PR DESCRIPTION
**Pull request by abc982627271**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 4 / 5 (80.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|2|3|66.67%|[959]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|2|2|100.00%|[]|

**Review by sduzh — DISMISSED** [review]

**Review by sduzh — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 10 / 49 (20.41%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|7|46|15.22%|[191, 614, 646, 647, 836, 838, 839, 840, 841, 842, 843, 844, 948, 949, 992, 993, 998, 999, 1042, 1043, 1044, 1047, 1048, 1052, 1057, 1058, 1059, 1060, 1063, 1064, 1065, 1066, 1070, 1071, 1072, 1077, 1078, 1082, 1083]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 1]
```diff
@@ -974,6 +1027,62 @@ public void readFields(DataInput in) throws IOException {
         }
     }
 
+    public static class RecyclePartitionInfoV2 extends RecyclePartitionInfo implements GsonPreProcessable,
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:1030]: We will support list partition recycle, and better to abstract `RecycleRangePartitionInfoV2` and `RecycleListPartitionInfoV2`.

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 2]
```diff
@@ -974,6 +1039,79 @@ public void readFields(DataInput in) throws IOException {
...
+
+        @Override
+        public void write(DataOutput out) throws IOException {
+            out.writeLong(-1L);
+            String json = GsonUtils.GSON.toJson(this);
+            Text.writeString(out, json);
+        }
+    }
+
+    public static class RecycleRangePartitionInfo extends RecyclePartitionInfoV2 implements GsonPreProcessable,
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:1070]: Need to add a new RuntimeTypeAdapterFactory in `GsonUtil` to indicate the inheritance relationship.

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 22 / 69 (31.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|14|61|22.95%|[191, 226, 604, 623, 650, 651, 653, 659, 660, 849, 852, 853, 855, 856, 857, 858, 961, 962, 1000, 1001, 1009, 1010, 1053, 1054, 1055, 1058, 1059, 1063, 1068, 1069, 1070, 1071, 1086, 1088, 1089, 1092, 1096, 1097, 1098, 1099, 1103, 1104, 1105, 1110, 1111, 1115, 1116]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|5|5|100.00%|[]|

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 29 / 69 (42.03%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|21|61|34.43%|[226, 604, 623, 650, 651, 653, 659, 660, 849, 852, 853, 855, 856, 857, 858, 961, 962, 1000, 1001, 1009, 1010, 1058, 1059, 1063, 1068, 1069, 1070, 1071, 1092, 1096, 1097, 1098, 1099, 1103, 1104, 1105, 1110, 1111, 1115, 1116]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|5|5|100.00%|[]|

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 33 / 69 (47.83%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|25|61|40.98%|[226, 650, 651, 653, 659, 660, 849, 855, 856, 858, 859, 860, 861, 964, 965, 1003, 1004, 1012, 1013, 1061, 1062, 1071, 1072, 1073, 1074, 1099, 1100, 1101, 1102, 1106, 1107, 1108, 1113, 1114, 1118, 1119]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 50 / 69 (72.46%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|42|61|68.85%|[226, 650, 651, 653, 659, 660, 849, 855, 856, 858, 859, 860, 861, 964, 965, 1003, 1004, 1012, 1013]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|5|5|100.00%|[]|

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 54 / 69 (78.26%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|46|61|75.41%|[226, 650, 651, 653, 659, 660, 849, 855, 856, 858, 859, 860, 861, 1003, 1004]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|5|5|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 3]
```diff
@@ -950,6 +999,29 @@ public boolean isInMemory() {
...
         }
 
+        public void setDbId(long dbId) {
+            this.dbId = dbId;
+        }
+    }
+
+    // only for RangePartition
+    public static class RecyclePartitionInfoV1 extends RecyclePartitionInfo {
+        protected Range<PartitionKey> range;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:1009]:
```suggestion
        private Range<PartitionKey> range;
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 4]
```diff
@@ -974,6 +1046,79 @@ public void readFields(DataInput in) throws IOException {
...
+            super(dbId, tableId, partition, dataProperty, replicationNum,
+                    isInMemory, storageInfo);
+            this.range = range;
+        }
+
+        public Range<PartitionKey> getRange() {
+            return range;
+        }
+
+        byte[] serializeRange(Range<PartitionKey> range) throws IOException {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:1098]:
```suggestion
        private byte[] serializeRange(Range<PartitionKey> range) throws IOException {
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 5]
```diff
@@ -974,6 +1046,79 @@ public void readFields(DataInput in) throws IOException {
...
+            ByteArrayOutputStream stream = new ByteArrayOutputStream();
+            DataOutputStream dos = new DataOutputStream(stream);
+            RangeUtils.writeRange(dos, range);
+            return stream.toByteArray();
+        }
+
+        Range<PartitionKey> deserializeRange(byte[] serializedRange) throws IOException {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:1105]:
```suggestion
        private Range<PartitionKey> deserializeRange(byte[] serializedRange) throws IOException {
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 6]
```diff
@@ -974,6 +1046,79 @@ public void readFields(DataInput in) throws IOException {
         }
     }
 
+    public static class RecyclePartitionInfoV2 extends RecyclePartitionInfo  {
+        @SerializedName(value = "storageInfo")
+        protected StorageInfo storageInfo;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:1051]:
```suggestion
        private StorageInfo storageInfo;
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 7]
```diff
@@ -974,6 +1046,79 @@ public void readFields(DataInput in) throws IOException {
...
+            out.writeLong(-1L);
+            String json = GsonUtils.GSON.toJson(this);
+            Text.writeString(out, json);
+        }
+    }
+
+    public static class RecycleRangePartitionInfo extends RecyclePartitionInfoV2 implements GsonPreProcessable,
+            GsonPostProcessable {
+
+        protected Range<PartitionKey> range;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:1080]:
```suggestion
        private Range<PartitionKey> range;
```

**Review by wyb — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Pull request merged**